File: web/lib/itunes.js
/**
 * Public App Store catalogue client for the browser.
 *
 * Talks to Apple directly by default (`LOOKUP_URL`/`SEARCH_URL` below), the
 * only mode a fresh clone with no backend configured ever has — and every
 * lookup this makes is public catalogue data, never anything about the
 * person running it. A deployment that *has* configured `community/`
 * instead routes every instance of this client through that Worker's
 * `/itunes/*` relay (see `itunesRelayOptions` in `lib/community.js`,
 * wired in by both `app.js` and `landing.js`) — a server-to-server fetch
 * has no CORS story to break, unlike a direct request to Apple's
 * undocumented endpoint, whose answer on that front has flip-flopped more
 * than once during this project.
 *
 * It is still somebody else's free service either way. This client is as
 * polite as the Python one is from a terminal: responses are cached,
 * requests are spaced, and nothing is fetched twice within a session.
 */

export const SEARCH_URL = 'https://itunes.apple.com/search';
export const LOOKUP_URL = 'https://itunes.apple.com/lookup';

/**
 * The screenshot relay's URL, once you have deployed `worker/` — see its
 * README. `null` disables the fallback entirely: the page then behaves
 * exactly as it did before, reporting withheld screenshots as "not exposed"
 * rather than fetching them from anywhere.
 *
 * A separate Worker from the `community/` one `itunesRelayOptions` points
 * at — this one only ever receives a track id and a country code, both
 * already public, and exists purely to recover screenshots the catalogue
 * response withholds. Set this only once you have read `worker/README.md`
 * and are comfortable with what it does.
 */
export const SCREENSHOT_RELAY_URL = 'https://launchpilot-screenshot-relay.kaizenapp-contact.workers.dev';

/** Apple publishes no rate limit; ~20/minute is the understood ceiling. */
export const MIN_REQUEST_INTERVAL_MS = 1200;

const CACHE_TTL_MS = 12 * 60 * 60 * 1000;
const CACHE_PREFIX = 'launchpilot:itunes:';

/** Page sizes to fall back through when the full response cannot be fetched. */
const FALLBACK_LIMITS = [150, 100, 50];

export class MarketDataError extends Error {}

const sleep = (ms) => new Promise((resolve) => setTimeout(resolve, ms));

/**
 * Session-scoped response cache.
 *
 * sessionStorage rather than localStorage: this is other people's catalogue
 * data, useful for the length of a working session and stale by tomorrow.
 * Keeping it out of long-term storage also means closing the tab leaves
 * nothing behind.
 */
class ResponseCache {
  constructor(storage = globalThis.sessionStorage) {
    this.storage = storage;
  }

  get(key) {
    if (!this.storage) return null;
    try {
      const raw = this.storage.getItem(CACHE_PREFIX + key);
      if (!raw) return null;
      const { at, payload } = JSON.parse(raw);
      if (Date.now() - at > CACHE_TTL_MS) return null;
      return payload;
    } catch {
      return null;
    }
  }

  set(key, payload) {
    if (!this.storage) return;
    try {
      this.storage.setItem(CACHE_PREFIX + key, JSON.stringify({ at: Date.now(), payload }));
    } catch {
      // Quota exceeded, or storage disabled. A cache that cannot write is a
      // slow cache, not a broken program — drop the oldest entries and move on.
      this.evictOldest();
    }
  }

  evictOldest() {
    if (!this.storage) return;
    try {
      const keys = Object.keys(this.storage).filter((k) => k.startsWith(CACHE_PREFIX));
      for (const key of keys.slice(0, Math.ceil(keys.length / 2))) {
        this.storage.removeItem(key);
      }
    } catch {
      /* nothing further to try */
    }
  }

  clear() {
    if (!this.storage) return;
    try {
      for (const key of Object.keys(this.storage)) {
        if (key.startsWith(CACHE_PREFIX)) this.storage.removeItem(key);
      }
    } catch {
      /* nothing further to try */
    }
  }
}

export class ITunesClient {
  /**
   * `lookupUrl`/`searchUrl` default to Apple directly — what a fresh clone
   * with no `community/` backend deployed gets, since there is nowhere
   * else to route through. Every caller in this app (`app.js`,
   * `landing.js`) overrides them via `itunesRelayOptions()` in
   * `lib/community.js` once a backend *is* configured, so catalogue
   * lookups go through that Worker's `/itunes/lookup` and `/itunes/search`
   * instead — a server-to-server fetch has no CORS story to break, unlike
   * a request straight from the browser to Apple's undocumented endpoint.
   *
   * `searchFallbackUrl`/`lookupFallbackUrl` only matter when `searchUrl`/
   * `lookupUrl` point at the relay: Apple rate-limits its endpoints by
   * source IP, and the relay's IP is Cloudflare's own shared Workers egress
   * range — shared with every other customer's Workers, not just this one.
   * That range has been observed getting rate-limited (and, on `/lookup`,
   * returning a transient 5xx) for stretches longer than a single retry
   * window, while direct-from-browser calls kept working the whole time. So
   * a failed relay call retries once, straight against Apple — trading back
   * a little of the CORS flakiness the relay exists to avoid, only when the
   * relay itself is the thing failing.
   */
  constructor({
    cache = new ResponseCache(),
    minInterval = MIN_REQUEST_INTERVAL_MS,
    fetchImpl,
    screenshotRelayUrl = SCREENSHOT_RELAY_URL,
    lookupUrl = LOOKUP_URL,
    lookupFallbackUrl = LOOKUP_URL,
    searchUrl = SEARCH_URL,
    searchFallbackUrl = SEARCH_URL,
  } = {}) {
    this.cache = cache;
    this.minInterval = minInterval;
    this.fetchImpl = fetchImpl ?? ((...args) => globalThis.fetch(...args));
    this.screenshotRelayUrl = screenshotRelayUrl;
    this.lookupUrl = lookupUrl;
    this.lookupFallbackUrl = lookupFallbackUrl;
    this.searchUrl = searchUrl;
    this.searchFallbackUrl = searchFallbackUrl;
    this.lastRequestAt = null;
  }

  async throttle() {
    if (this.lastRequestAt === null) return;
    const remaining = this.minInterval - (Date.now() - this.lastRequestAt);
    if (remaining > 0) await sleep(remaining);
  }

  async get(url, params, { subject = 'request' } = {}) {
    const query = new URLSearchParams(params).toString();
    const key = `${url}?${query}`;

    const cached = this.cache?.get(key);
    if (cached) return cached;

    await this.throttle();

    let response;
    try {
      response = await this.fetchImpl(`${url}?${query}`);
    } catch (err) {
      throw new MarketDataError(
        `Could not reach the App Store catalogue: ${err.message}. Check your connection and try again.`,
      );
    } finally {
      this.lastRequestAt = Date.now();
    }

    if (!response.ok) {
      throw new MarketDataError(
        `The App Store catalogue returned ${response.status} for "${subject}". ` +
          'This endpoint rate-limits; wait a minute and try again.',
      );
    }

    let payload;
    try {
      payload = await response.json();
    } catch {
      // The endpoint answers throttled requests with a non-JSON body and a 200.
      throw new MarketDataError(
        `The App Store catalogue returned an unreadable response for "${subject}". ` +
          'This usually means rate limiting; wait a minute and try again.',
      );
    }

    this.cache?.set(key, payload);
    return payload;
  }

  /**
   * Search, falling back to a smaller page if the full one cannot be fetched.
   *
   * A 200-result response is around 1.5 MB — the endpoint returns every field
   * it has, including full descriptions and device lists, with no way to ask
   * for less. That is fine on a desktop connection and not fine on a phone, a
   * throttled network, or behind a proxy that caps response bodies.
   *
   * Rather than quietly lowering the limit for everyone — which would change
   * the competitive-depth score and silently break parity with the CLI — the
   * full page is attempted first and the fallback is *reported*, so a reduced
   * sample is visible in the output instead of being mistaken for a thinner
   * market.
   */
  async search(term, { country = 'us', limit = 200 } = {}) {
    try {
      return await this.searchAt(this.searchUrl, term, { country, limit });
    } catch (err) {
      // No distinct fallback configured (fresh clone, direct-to-Apple
      // already) — nothing left to retry against.
      if (this.searchUrl === this.searchFallbackUrl) throw err;
      return await this.searchAt(this.searchFallbackUrl, term, { country, limit });
    }
  }

  async searchAt(url, term, { country, limit }) {
    const attempts = [limit, ...FALLBACK_LIMITS.filter((l) => l < limit)];
    let lastError = null;

    for (const attempt of attempts) {
      try {
        const payload = await this.get(
          url,
          { term, country: country.toLowerCase(), entity: 'software', limit: String(attempt) },
          { subject: term },
        );
        return {
          resultCount: payload.resultCount ?? 0,
          results: payload.results ?? [],
          limitUsed: attempt,
          limitRequested: limit,
          reduced: attempt < limit,
        };
      } catch (err) {
        lastError = err;
      }
    }
    throw lastError;
  }

  /**
   * Fetch one app by numeric App Store id or by bundle id.
   *
   * `country` is always sent. Whether Apple's lookup endpoint returns CORS
   * headers for `?id=X` alone versus `?id=X&country=us` has flipped more
   * than once during this project — sending both params is the safer bet
   * either way, and callers who can't afford to be at the mercy of that
   * (the landing page) route around it entirely via `lookupUrl`.
   */
  async lookup(appId, { country = 'us' } = {}) {
    const isNumericId = /^\d+$/.test(appId);
    const params = isNumericId
      ? { id: appId, country: country.toLowerCase() }
      : { bundleId: appId, country: country.toLowerCase() };

    try {
      const payload = await this.get(this.lookupUrl, params, { subject: appId });
      return payload.results?.[0] ?? null;
    } catch (err) {
      // No distinct fallback configured (fresh clone, direct-to-Apple
      // already) — nothing left to retry against.
      if (this.lookupUrl === this.lookupFallbackUrl) throw err;
      const payload = await this.get(this.lookupFallbackUrl, params, { subject: appId });
      return payload.results?.[0] ?? null;
    }
  }

  /**
   * Recover screenshots via the relay when the catalogue withheld them.
   *
   * Returns `null` when the relay is not configured, unreachable, or found
   * nothing — every case degrades to "not exposed", never an error, since a
   * developer whose relay is down should see the same page they saw before
   * one existed, not a broken tool.
   */
  async fetchPageScreenshots(trackId, { country = 'us' } = {}) {
    if (!this.screenshotRelayUrl) return null;

    const key = `${this.screenshotRelayUrl}/screenshots?id=${trackId}&country=${country.toLowerCase()}`;
    const cached = this.cache?.get(key);
    if (cached) return cached;

    const query = new URLSearchParams({ id: String(trackId), country: country.toLowerCase() });
    let response;
    try {
      response = await this.fetchImpl(`${this.screenshotRelayUrl}/screenshots?${query}`);
    } catch {
      return null;
    }
    if (!response.ok) return null;

    let payload;
    try {
      payload = await response.json();
    } catch {
      return null;
    }

    const iphone = Array.isArray(payload?.iphone) ? payload.iphone : [];
    const ipad = Array.isArray(payload?.ipad) ? payload.ipad : [];
    if (!iphone.length && !ipad.length) return null;

    const result = { iphone, ipad };
    this.cache?.set(key, result);
    return result;
  }
}

export { ResponseCache };

File: src/launchpilot/core/clients/itunes.py
"""Public App Store catalogue client.

Apple's iTunes Search endpoint is public, documented and needs no credentials,
which is the whole reason the niche analysis can work for everyone rather than
only for developers who already run Apple Search Ads campaigns.

It is also somebody else's free service, so this client is deliberately polite:
responses are cached on disk, requests are spaced by a minimum interval, and the
default page size is the one that answers the question rather than the maximum
the API allows.

Network access lives here and nowhere else. :mod:`market_analyzer` scores
snapshots it is handed, so the entire scoring path is testable offline.
"""

from __future__ import annotations

import contextlib
import datetime as dt
import hashlib
import json
import time
from pathlib import Path
from typing import Any, Protocol, runtime_checkable

import httpx

from launchpilot.core.errors import LaunchPilotError

SEARCH_URL = "https://itunes.apple.com/search"
LOOKUP_URL = "https://itunes.apple.com/lookup"

# Apple does not publish a rate limit for this endpoint; ~20 requests/minute is
# the community-understood ceiling. Stay well under it.
MIN_REQUEST_INTERVAL = 3.0

DEFAULT_CACHE_TTL = dt.timedelta(hours=12)

USER_AGENT = "launchpilot/0.1 (+https://github.com/pP4010/launchpilot)"


class MarketDataError(LaunchPilotError):
    """The catalogue could not be reached or returned something unusable."""


@runtime_checkable
class MarketDataSource(Protocol):
    """Where competitor listings come from.

    Implemented by :class:`ITunesSearchClient` today and by recorded fixtures in
    the tests. Keeping this a Protocol is what lets the analyser be exercised
    without touching the network.
    """

    def search(self, term: str, *, country: str, limit: int) -> tuple[int, list[dict[str, Any]]]:
        """Return ``(result_count, raw_entries)`` for a search term."""
        ...

    def lookup(self, app_id: str, *, country: str) -> dict[str, Any] | None:
        """Return one app's catalogue entry, by numeric track id or bundle id."""
        ...


class ResponseCache:
    """Small on-disk cache keyed by the request parameters.

    Re-running an analysis while tuning weights is the common case, and it should
    not re-hit Apple each time.
    """

    def __init__(self, directory: Path, ttl: dt.timedelta = DEFAULT_CACHE_TTL) -> None:
        self.directory = directory
        self.ttl = ttl

    def _path(self, key: str) -> Path:
        digest = hashlib.sha256(key.encode("utf-8")).hexdigest()[:32]
        return self.directory / f"{digest}.json"

    def get(self, key: str) -> dict[str, Any] | None:
        path = self._path(key)
        if not path.is_file():
            return None
        age = dt.datetime.now(dt.UTC) - dt.datetime.fromtimestamp(path.stat().st_mtime, dt.UTC)
        if age > self.ttl:
            return None
        try:
            data: dict[str, Any] = json.loads(path.read_text(encoding="utf-8"))
        except (OSError, json.JSONDecodeError):
            return None
        return data

    def set(self, key: str, payload: dict[str, Any]) -> None:
        self.directory.mkdir(parents=True, exist_ok=True)
        # A cache that cannot write is a slow cache, not a broken program.
        with contextlib.suppress(OSError):
            self._path(key).write_text(json.dumps(payload), encoding="utf-8")


class ITunesSearchClient:
    """Reads the public App Store catalogue over HTTP."""

    def __init__(
        self,
        *,
        cache: ResponseCache | None = None,
        timeout: float = 15.0,
        min_interval: float = MIN_REQUEST_INTERVAL,
        client: httpx.Client | None = None,
    ) -> None:
        self._cache = cache
        self._timeout = timeout
        self._min_interval = min_interval
        self._client = client
        self._last_request_at: float | None = None

    def _throttle(self) -> None:
        if self._last_request_at is None:
            return
        elapsed = time.monotonic() - self._last_request_at
        remaining = self._min_interval - elapsed
        if remaining > 0:
            time.sleep(remaining)

    def _get(self, url: str, params: dict[str, str], *, subject: str) -> dict[str, Any]:
        """One cached, throttled GET with catalogue errors translated.

        ``subject`` only appears in error messages, so a failure names the term
        or app the user asked about rather than a URL.
        """
        cache_key = json.dumps({"url": url, **params}, sort_keys=True)

        if self._cache is not None:
            cached = self._cache.get(cache_key)
            if cached is not None:
                return cached

        self._throttle()
        try:
            if self._client is not None:
                response = self._client.get(url, params=params)
            else:
                headers = {"User-Agent": USER_AGENT}
                response = httpx.get(url, params=params, timeout=self._timeout, headers=headers)
            self._last_request_at = time.monotonic()
            response.raise_for_status()
        except httpx.HTTPStatusError as exc:
            raise MarketDataError(
                f"App Store catalogue returned {exc.response.status_code} for {subject!r}. "
                "This endpoint rate-limits; wait a minute and retry."
            ) from exc
        except httpx.HTTPError as exc:
            raise MarketDataError(f"Could not reach the App Store catalogue: {exc}") from exc

        try:
            payload: dict[str, Any] = response.json()
        except ValueError as exc:
            # The endpoint answers rate-limited requests with a non-JSON body.
            raise MarketDataError(
                f"App Store catalogue returned a non-JSON response for {subject!r}. "
                "This usually means rate limiting; wait a minute and retry."
            ) from exc

        if self._cache is not None:
            self._cache.set(cache_key, payload)
        return payload

    def search(self, term: str, *, country: str, limit: int) -> tuple[int, list[dict[str, Any]]]:
        payload = self._get(
            SEARCH_URL,
            {
                "term": term,
                "country": country.lower(),
                "entity": "software",
                "limit": str(limit),
            },
            subject=term,
        )
        return int(payload.get("resultCount", 0)), list(payload.get("results", []))

    def lookup(self, app_id: str, *, country: str) -> dict[str, Any] | None:
        """Fetch one app by numeric track id or by bundle id.

        The endpoint takes different parameter names for the two, and returns an
        empty result set rather than a 404 when nothing matches.

        ``country`` is omitted for numeric ids, matching the browser client. A
        track id already identifies one app across every storefront, and the
        parameter only varies pricing fields nothing here reads — but combining
        the two makes Apple drop its CORS headers, which breaks the web build.
        Keeping the request shapes identical keeps the two clients honest.
        """
        params = (
            {"id": app_id} if app_id.isdigit() else {"bundleId": app_id, "country": country.lower()}
        )
        payload = self._get(LOOKUP_URL, params, subject=app_id)
        results = payload.get("results") or []
        return dict(results[0]) if results else None


def default_cache_dir() -> Path:
    return Path.home() / ".cache" / "launchpilot" / "itunes"
